src/main/java/com/google/sps/servlets/PotentialMatchesServlet.java

```diff
@@ -60,28 +70,69 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
   }
 
   /**
-  * Loads the next potential match for a user
+  * Loads the match information for a specified user if match information is not
+  * already in datastore.
   *
-  * <p>Currently is just loading all the possible matches and obtaining one of them.
+  * <p>Creates new match information entity for user if it had not already been stored.
   *
-  * <p>TODO(#21): Obtain potential match results from datastore instead of running the 
-  * potential match finding methods repeatedly.
+  * @param userID The ID of the user who's match information is being checked/loaded into datastore
+  */
+  private void loadUserMatchInformation(String userID) {
+    Entity matchInfoEntity = datastore.prepare(new Query(MATCH_INFO_ENTITY).setFilter(
+      new FilterPredicate(USER_ID_PROPERTY, FilterOperator.EQUAL, userID))).asSingleEntity();
+    
+    if (matchInfoEntity == null) {
+      addMatchInfoToDatastore(userID);
+    }
+  }
+  
+  /**
+  * Given a specific user, get their next potential match for their feed page
+  *
+  * @param userID The user whose potential match is being retrieved
+  */
+  private String getNextPotentialMatchID(String userID) {
+    Entity matchInfoEntity = datastore.prepare(new Query(MATCH_INFO_ENTITY).setFilter(
+      new FilterPredicate(USER_ID_PROPERTY, FilterOperator.EQUAL, userID))).asSingleEntity();
+
+    List<String> potentialMatches = (ArrayList<String>) matchInfoEntity.getProperty(POTENTIAL_MATCHES_PROPERTY);
+
+    String nextPotentialMatchID;
+    
+    if (potentialMatches != null && potentialMatches.iterator().hasNext()) {
```

Comment:
nit: instead of using an Iterator here, it may be more clear to just deal with the 1st element specifically, especially since the way the code works currently is a bit unclear -- i.e. does calling `remove()` on the iterator affect the original Entity or not? (that's actually dependent on the implementation of `getProperty` -- whether it copies its internal representation or returns it directly)

e.g. could rewrite as

```java
if (potentialMatches != null && !potentialMatches.isEmpty()) {
  nextPotentialMatchID = potentialMatches.get(0);
  matchIntoEntity.setProperty(POTETNIAL_MATCHES_PROPERTY, potentialMatches.subList(1, potentialMatches.size()));
} else {
  // ...
}
```

This also avoids any assumption that the return value of `getProperty` is mutable -- e.g. if `getProperty` returns an `ImmutableList`, calling `iterator.remove()` would throw.

Reply:
That makes sense. I changed this to include your rewrite suggestion!